update docs with out-of-service taint information

clobrano opened the pull request:
<!-- Thanks for contributing to our project! We appreciate your time and effort.
Please fill out the information below to expedite the review and merge of your pull request.
-->

#### Why we need this PR
<!--
Outline the purpose of this PR, whether it's addressing a bug, implementing a new feature, or solving a specific problem.
-->
The documentation only contains information about the "deletion strategy", while also `out-of-service` taint is supported now.

#### Changes made
<!-- Outline the specific changes made in this merge request. -->
Add out-of-service taint support information



#### Which issue(s) this PR fixes
<!--
Any reference to relevant issue(s).
Please use the following format, so that the issue will be automatically closed when this PR is merged (see https://help.github.com/articles/closing-issues-using-keywords/)

`Fixes #<issue number>`

If there is not a correspondent issue yet, you might want to open a new one yourself, describing the problem you observed.
-->


#### Test plan
<!--
Please, make sure that this PR meets all the necessary quality gates before submitting for review:

- Existing Unit and E2E tests are passing
- New features or bug fixes should be covered by new Unit and/or E2E tests.

This will help us to ensure that your changes are working as expected and will not break in the future.
 
In order to save cloud resources, we invite you to submit the PR as a draft and run a single E2E test job, e.g. adding a comment to the PR with the following message in order to run E2E test on OCP 4.15 only:

> /test 4.15-openshift-e2e

In case you are unable to verify E2E test prior to submit the PR, we suggest to use the WIP (Work In Progress) title prefix (e.g. "[WIP] <Title of the PR>"), and then follow the above mentioned manual test steps. Once the E2E job passes, you can remove the WIP prefix and request a review.
-->

clobrano committed:
Fix typo

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

clobrano committed:
Add out-of-service taint support information

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

mshitrit commented on recover_affected_workloads.md:
```diff
@@ -21,12 +21,17 @@ For example, in the case where the node fails to update its status to the
-deletion][2]
+control plane needs to know that the running node is gone and there are two ways to do it:
+
+- deleting the node object, which will trigger [scheduling the pods for deletion][2]
```
> deleting the node object

Good job on updating the doc, but I think we are confusing here couple of different strategies.
- Node deletion (which was completely removed), this strategy used to delete the Node.
- ResourceDeletion - which explicitly deletes the pods of the remediated node
- OutOfServiceTaint - which implicitly causes the node's resource removal in a more gracefully way by placing a taint on the node.

ATM only the later 2 are being used.
SNR also has a third strategy`Automatic`  which basically let's the let's the remediator choose how to perform the remediation, which is something which might be worth mentioning as well.

"Under the hood" `Automatic`   just dynamically chooses OutOfServiceTaint strategy in case it's supported, otherwise it chooses ResourceDeletion, but I don't think we want to document this part.

razo7 replied:
I'm not sure `Automatic` is relevant here. IMHO it could be mentioned in a different subsection or as a note on remediation strategy that is supported in  SNR (and in the future in Medik8s when FAR will support it as well).

razo7 commented on recover_affected_workloads.md:
```diff
@@ -21,12 +21,17 @@ For example, in the case where the node fails to update its status to the
+control plane needs to know that the running node is gone and there are two ways to do it:
+
+- deleting the node object, which will trigger [scheduling the pods for deletion][2]
+- adding an [`out-of-service` taint][4] on the Node, which triggers pods on the node
```
The link number should be 3 as it right after 2.

clobrano replied:
I think it would be just easier to use implicit reference links

razo7 replied:
SGTM I can't recall why it was used this way. NP for changing it from my POV as long we are consistent with the rest of the website pages.

clobrano replied:
the other pages can be modified in followup PR

razo7 commented on recover_affected_workloads.md:
```diff
@@ -61,3 +66,4 @@ TBD
 [1]: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/
 [2]: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection
 [3]: https://libvirt.org/formatdomain.html#watchdog-device
+[4]: https://kubernetes.io/docs/reference/labels-annotations-taints/#node-kubernetes-io-out-of-service
```
I think their order should be switched because of my [above comment](https://github.com/medik8s/docs/pull/70/files#r1597584040)

razo7 commented on recover_affected_workloads.md:
```diff
@@ -21,12 +21,17 @@ For example, in the case where the node fails to update its status to the
+- adding an [`out-of-service` taint][4] on the Node, which triggers pods on the node
+to be forcefully deleted if there are no matching tolerations on the pods.
+Persistent volumes attached to the shutdown node will be detached, and new pods
+will be created successfully on a different running node. This is the "right" way to do it.
```
```suggestion
will be created successfully on a different running node. This is the recommended way to do it.
```

razo7 commented on recover_affected_workloads.md:
```diff
@@ -21,12 +21,17 @@ For example, in the case where the node fails to update its status to the
+Persistent volumes attached to the shutdown node will be detached, and new pods
+will be created successfully on a different running node. This is the "right" way to do it.
+
+Currently, the "deletion strategy" is still supported by our [fence-agents-remediation](https://github.com/medik8s/fence-agents-remediation)
+operator (FAR), while both FAR and [self-node-remediation](/remediation/self-node-remediation/self-node-remediation/) (SNR) operator support the `out-of-service` taint.
```
Over here it would be nice to add the automatic strategy while changing `the "deletion strategy"` -> `the "resource deletion strategy"`

razo7 commented on recover_affected_workloads.md:
```diff
@@ -21,12 +21,17 @@ For example, in the case where the node fails to update its status to the
-control plane needs to know that the running node is gone and the only way to achieve
-that is by deleting the node object, which will trigger [scheduling the pods for
-deletion][2]
+control plane needs to know that the running node is gone and there are two ways to do it:
```
```suggestion
control plane needs to know that the running node is gone and there are two ways for Non-graceful node shutdown handling[2]:
```

[2] https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown

slintes replied:
this suggested change does not make sense, "non-graceful node shutdown" is the name of the feature which uses the out of service taint for dealing with nodes which are shut down but the control plane not knowing it yet

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,29 @@ nav_order: 8
-that is by deleting the node object, which will trigger [scheduling the pods for
-deletion][2]
+In order to gain the workload back, reschedule it and start it all over, the control plane
+needs to know that the running node is gone and there are two strategies that accomplish
```
> `that the running node is gone`

well, the node resource still exists.
maybe: `that the workloads's pod is not running anymore`?

and may start a new sentence afterwards: `... anymore. There are two...`

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,29 @@ nav_order: 8
+needs to know that the running node is gone and there are two strategies that accomplish
+that:
+
+- `ResourceDeletion`, which will remove the pods on the node (rather than removing the
```
I would not mention node deletion at all anymore... 🤔

clobrano replied:
However, it is still an option. I can rephrase the whole thing making clearer that `ResourceDeletion` not-suggested/old option, instead than suggesting `OutOfServiceTaint`

slintes replied:
I mean NODE deletion, that latter part of the list item ("rather than...")

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,29 @@ nav_order: 8
+
+- `ResourceDeletion`, which will remove the pods on the node (rather than removing the
+node object) recovering the workloads faster.
+- [`OutOfServiceTaint`][], which triggers pods on the node to be forcefully deleted if
```
maybe add somehow that this is a new core k8s feature named non graceful node shutdown and link to this blog? https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/

clobrano committed:
Rephrase Strategies description

- clarify that is the workload not running anymore, more than the Node
- add that OutOfServiceTaint is a new core k8s feature with a link to
  more information
- add that ResourceDeletion is an old/unsuggested strategy

mshitrit commented:
/lgtm
/hold
giving others chance to review, feel free to unhold

razo7 commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,31 @@ nav_order: 8
+  be detached, and new pods will be created successfully on a different running node. This
+  is the recommended strategy.
+
+- `ResourceDeletion`, which will remove the pods on the node. This strategy is still supported for
```
It seems odd to me that the bullet begins with a comma (`,`). Maybe a hyphen/colon (`-/:`) would seem better here.

razo7 replied:
Looking on other pages under Medik8s website, and we usually use hyphens:

- https://www.medik8s.io/about_us/
- https://www.medik8s.io/remediation/remediation/
- https://www.medik8s.io/remediation/fence-agents-remediation/fence-agents-remediation/

clobrano replied:
I'll also rephrase this one like FAR readme (s/remove/delete/)

clobrano committed:
Code review changes on recover_affected_workloads

- Clarified description of remediation strategies
- Fixed typo

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,31 @@ nav_order: 8
 In kubernetes world there is no concept of rescheduling. When pods are bounded to a
 node they run till termination. In a case of hardware failure, or some other
-resource problem that hang the OS, there is no built-in solution to migrate or
+resource problem that hangs the OS, there is no built-in solution to migrate or
```
s/hangs/freezes?

clobrano replied:
quick search: it seems that `hangs` is more used when the problem _can_ be restored. Change accepted :)

razo7 commented:
/lgtm

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,31 @@ nav_order: 8
-that is by deleting the node object, which will trigger [scheduling the pods for
-deletion][2]
+In order to gain the workload back, reschedule it and start it all over, the control plane
+needs to know that the the workloads's pod is not running anymore. There are two
```
I think it should either be `workload's pod` or `workloads' pods` 🤔 
Since we use singular workload in the line above, the former fits better

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,31 @@ nav_order: 8
+- [`OutOfServiceTaint`][] - new core Kubernetes feature named [Non-Graceful Node
+  Shutdown], it triggers pods on the node to be forcefully deleted if there are no
+  matching tolerations on the pods. Persistent volumes attached to the shutdown node will
+  be detached, and new pods will be created successfully on a different running node. This
```
how do you know that pod creation will be successful?

clobrano replied:
good point

slintes commented on recover_affected_workloads.md:
```diff
@@ -8,25 +8,31 @@ nav_order: 8
+  be detached, and new pods will be created successfully on a different running node. This
+  is the recommended strategy.
+
+- `ResourceDeletion` - it deletes the pods and detaches associated volumes on the node. This strategy
```
this sounds like detaching volumes is something we explicitly do, but we don't (anymore)...?

slintes commented:
/approve 
/hold cancel

slintes commented:
/approve

slintes commented:
oh, we don't have a OWNERS file 😎

slintes commented:
https://github.com/medik8s/docs/pull/71

slintes commented:
/approve
